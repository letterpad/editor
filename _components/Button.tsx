import React, { SFC, MouseEventHandler } from "react";
import classnames from "classnames";
import styled from "styled-components";

const Wrapper = styled.span`
  ${(p: any) => p.styleString}
`;
const TextIcon = styled.span`
  border-radius: 50%;
  border: 1px solid;
  font-size: 12px;
  font-weight: 600;
  width: 32px;
  height: 32px;
  vertical-align: text-bottom;
  margin-right: 6px;
  display: inline-block;
  padding: 3px;
  font-family: sans-serif;
  line-height: 25px;
`;

interface ButtonProps {
  onMouseDown: MouseEventHandler;
  icon?: string;
  isActive?: boolean;
  styleString?: string;
  iconText?: string;
}

const Button: SFC<ButtonProps> = ({
  onMouseDown,
  isActive,
  icon,
  iconText,
  styleString
}) => {
  const classes = classnames("button", {
    active: isActive
  });
  const isImageLink = icon && icon.indexOf(".") > 0;
  const isMaterial = !isImageLink && !iconText;
  return (
    <Wrapper
      styleString={styleString}
      className={classes}
      onMouseDown={onMouseDown}
      icon={icon}
    >
      {isImageLink && <span className="custom-icons" />}
      {iconText && <TextIcon className="lp-text-icon">{iconText}</TextIcon>}
      {isMaterial && <span className="material-icons">{icon}</span>}
    </Wrapper>
  );
};

export default Button;
