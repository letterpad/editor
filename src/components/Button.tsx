import React, { SFC, MouseEventHandler } from "react";
import classnames from "classnames";
import styled from "styled-components";

const Wrapper = styled.span`
  ${(p: any) => p.styleString}
`;
interface ButtonProps {
  onMouseDown: MouseEventHandler;
  icon: string;
  isActive?: boolean;
  styleString?: string;
}

const Button: SFC<ButtonProps> = ({
  onMouseDown,
  isActive,
  icon,
  styleString
}) => {
  const classes = classnames("button", {
    active: isActive
  });
  const isCustomIcon = icon.indexOf(".") > 0;
  return (
    <Wrapper
      styleString={styleString}
      className={classes}
      onMouseDown={onMouseDown}
      icon={icon}
    >
      {isCustomIcon ? (
        <span className="custom-icons" />
      ) : (
        <span className="material-icons">{icon}</span>
      )}
    </Wrapper>
  );
};

export default Button;
