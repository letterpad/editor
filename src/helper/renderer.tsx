import React, { ReactNode } from "react";
import { PluginsMap } from "../plugins";
import { Plugin } from "slate-react";

export type OriginalRenderNodeProps = Parameters<
  Required<Plugin>["renderNode"]
>;

export interface RenderNodeHandler {
  (params: {
    props: OriginalRenderNodeProps[0];
    next?: OriginalRenderNodeProps[2];
    callbacks?: {
      [key: string]: any;
    };
    pluginsMap: PluginsMap;
  }): ReactNode;
}

// Search from the pluginsMap and give back the node to render
export const renderNode: RenderNodeHandler = ({
  props,
  next,
  callbacks,
  pluginsMap
}) => {
  if (pluginsMap.node[props.node.type]) {
    const RenderNode = pluginsMap.node[props.node.type].plugin.render;
    if (callbacks && callbacks.onBeforeRender) {
      callbacks.onBeforeRender({
        renderType: "node",
        type: props.node.type,
        props
      });
    }
    const noop = () => {};
    return <RenderNode {...props} next={next || noop} />;
  }
  return next && next();
};

export type OriginalRenderMarkProps = Parameters<
  Required<Plugin>["renderMark"]
>;

export interface RenderMarkHandler {
  (params: {
    props: OriginalRenderMarkProps[0];
    next?: OriginalRenderMarkProps[2];
    callbacks?: {
      [key: string]: any;
    };
    pluginsMap: PluginsMap;
  }): ReactNode;
}

// Search from the pluginsMap and give back the mark to render
export const renderMark: RenderMarkHandler = ({
  props,
  next,
  callbacks,
  pluginsMap
}) => {
  if (pluginsMap.mark[props.mark.type]) {
    const RenderMark = pluginsMap.mark[props.mark.type].plugin.render;
    if (callbacks && callbacks.onBeforeRender) {
      const onBeforeRender = callbacks.onBeforeRender({
        renderType: "mark",
        type: props.mark.type,
        props
      });
      if (onBeforeRender) return onBeforeRender;
    }
    return <RenderMark {...props} next={next} />;
  }
  return next && next();
};
