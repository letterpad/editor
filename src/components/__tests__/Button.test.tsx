import renderer from "react-test-renderer";
import React from "react";
import Button from "../Button";

describe("components", () => {
  describe("Button", () => {
    test("snapshot", () => {
      const onMouseDown = jest.fn();
      const component = renderer.create(<Button onMouseDown={onMouseDown} />);

      const tree: any = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
