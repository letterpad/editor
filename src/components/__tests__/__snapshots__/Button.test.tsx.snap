// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Button snapshot 1`] = `
<span
  className="button"
  onMouseDown={[MockFunction]}
>
  <span
    className="material-icons"
  />
</span>
`;
